### PR TITLE
Development

### PR DESCRIPTION
**Pull request by TobiasThrebank**
Hopefully it is better this time.

### CODE REVIEW

File: Unittest/Persons.addMultiple.test.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  let input = [
+    new Person('Zachariah', 'Portugal'),
+    new Person('Lue', 'Nehls'),
+    new Person('Crista', 'Risko'),
+    new Person('Denna', 'Giefer'),
+    new Person('Brandee', 'Hunsberger')
+  ];
+  let tmp = new Persons();
+
+  tmp.add(input);
```
Comment (fetis) [inline Unittest/Persons.addMultiple.test.ts]: edge case example might look like this.

```
const persons = new Persons();

tmp.add(null)

expect(persons).toEqual([]);
```

So we test not only the happy-path where we have the data, but and the situation when we don't.

File: Unittest/Persons.sort.test.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  ];
+  let expected = [
+    new Person('Denna', 'Giefer'),
+    new Person('Brandee', 'Hunsberger'),
+    new Person('Lue', 'Nehls'),
+    new Person('Zachariah', 'Portugal'),
+    new Person('Crista', 'Risko')
+  ];
+  let tmp = new Persons();
+  let f = function(collection) {
```
Comment (fetis) [inline Unittest/Persons.sort.test.ts]: >I don't understand it. If the method is supposed to take a function as argument, I have to supply one during the test, no?

Ok, let's define what we're trying to test here: the function call or function itself? I would say the both.
At the first,  we test that passed function is called with some expected arguments. For this we don't need real function, we can use [Jasmine spies](https://jasmine.github.io/tutorials/your_first_suite#section-Spies)
The second thing, if we provide sorting function as part of our library, then we need to move to another file and to test separately.

File: Unittest/Person.compare.test.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { Person } from './Person';
+
+it('should compare instances of Person', () => {
+  let p1 = new Person('Denna', 'Giefer');
+  let p2 = new Person('Brandee', 'Hunsberger');
+  let p3 = new Person('Denna', 'Giefer');
+
+  expect(p1.compare(p2)).toBe(false);
```
Comment (fetis) [inline Unittest/Person.compare.test.ts]: need to be split into 2 cases: equal, not equal

File: Unittest/Person.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+export class Person {
+
+  constructor(public firstName: string, public lastName: string) {
```
Comment (fetis) [inline Unittest/Person.ts]: we place every argument of constructor on a new line. this increases readability

```
constructor(
  public firstName: string, 
  public lastName: string
) {
```

File: Unittest/Persons.addSingle.test.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { Person } from './Person';
+import { Persons } from './Persons';
+
+it('should add single Person to the Persons collection', () => {
+  let tmp = new Persons();
+  let expected = [ new Person('Denna', 'Giefer') ];
```
Comment (fetis) [inline Unittest/Persons.addSingle.test.ts]: hm, normally you test samples are simpler than your class. You would try with `[{firstName: 'Denna', lastName: 'Giefer'}]`. Jasmine will compare corresponding fields for us. 
Otherwise, if you did a mistake (a `fristName` typo for example) you'll just repeat it in the test.

File: Unittest/Persons.remove.test.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+it('should remove a Person from Persons collection', () => {
+  let input = [
+    new Person('Zachariah', 'Portugal'),
+    new Person('Lue', 'Nehls'),
+    new Person('Crista', 'Risko'),
+    new Person('Denna', 'Giefer'),
+    new Person('Brandee', 'Hunsberger')
+  ];
+  let expected = [
```
Comment (fetis) [inline Unittest/Persons.remove.test.ts]: the same comment as I have above, a simple array with object literal will be enough `[{}, {}]`

File: Unittest/Persons.remove.test.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    new Person('Denna', 'Giefer'),
+    new Person('Brandee', 'Hunsberger')
+  ];
+  let tmp = new Persons();
+
+  tmp.add(input);
+
+  expect(tmp.remove(new Person('Crista', 'Risko'))).toBe(true);
+  expect(tmp.items).toEqual(expected);
+  expect(tmp.items[3]).toEqual(new Person('Brandee', 'Hunsberger'));
```
Comment (fetis) [inline Unittest/Persons.remove.test.ts]: this's not necessary, you checked it on previous line

File: Unittest/Persons.remove.test.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    new Person('Zachariah', 'Portugal'),
+    new Person('Lue', 'Nehls'),
+    new Person('Denna', 'Giefer'),
+    new Person('Brandee', 'Hunsberger')
+  ];
+  let tmp = new Persons();
+
+  tmp.add(input);
+
+  expect(tmp.remove(new Person('Crista', 'Risko'))).toBe(true);
```
Comment (fetis) [inline Unittest/Persons.remove.test.ts]: create a new instance to delete something? you either delete by reference or by some criteria(-s) like ID, name, etc. You need to accept a first and last name, not an object

You also don't check `false` value from `.remove`, i.e. when you don't find anything

File: Unittest/Persons.sort.test.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      }
+    }
+  }
+
+  tmp.add(input);
+  tmp.sortCollection();
+
+  expect(tmp.items).toEqual(expected);
+
+  tmp = new Persons();
```
Comment (fetis) [inline Unittest/Persons.sort.test.ts]: this needs to be split, since you test internal and external sorting in the same test case

File: Unittest/Persons.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        }
+      }
+      n = newn;
+    } while(n > 1);
+  }
+
+  public sortCollection(sortAlgorithm?: SortFunction) {
+    if(typeof sortAlgorithm === 'undefined') {
+      this.bubbleSort();
+    }
```
Comment (fetis) [inline Unittest/Persons.ts]: `} else {`

File: Unittest/Persons.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    for(let i = 0; i < this.collection.length; i++) {
+      if(this.collection[i].compare(person)) {
+        this.collection.splice(i, 1);
+        isFound = true;
+      }
+    }
+    return isFound;
+  }
+
+  private bubbleSort(): void {
```
Comment (fetis) [inline Unittest/Persons.ts]: it should goes after all static/public methods

File: Unittest/Persons.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
+import { Person } from './Person';
+
+type SortFunction = (_: Person[]) => void;
+
+export class Persons {
+  private collection: Person[] = [];
```
[the remark this answers is not in the document]
Comment (TobiasThrebank) [reply Unittest/Persons.ts]: So the generel structure would be
constructor
public properties
public methods
private properties
private methods
?

File: Unittest/Persons.addMultiple.test.ts (hunk 1 above)
Comment (TobiasThrebank) [reply Unittest/Persons.addMultiple.test.ts]: Understood, I didn't create edge cases for sorting though, because I would have to rewrite it to work with them. Please let me know if you want me to do that as well.

**Comment by TobiasThrebank** [comment]
I am happy to rewrite it until I get it right. I am not at all impatient about it. :)

File: Unittest/Person.ts [hunk 13]
```diff
@@ -1,8 +1,8 @@
 export class Person {
 
-  constructor(public firstName: string, public lastName: string) {
-
-  }
+  constructor(
+    public firstName: string,
+    public lastName: string) { }
```
Comment (fetis) [inline Unittest/Person.ts]:
```
constructor(
  public arg: type,
  private arg: type
) {
}
```

closing bracket on the new line

File: Unittest/Person.spec.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+  it('should create Person object', () => {
+    let tmp = new Person('Denna', 'Giefer');
+
+    expect(tmp.firstName).toBe('Denna');
+    expect(tmp.lastName).toBe('Giefer');
+  });
+
+  it('should create Person object with null argument', () => {
+    let tmp = new Person(null, 'Giefer');
```
Comment (fetis) [inline Unittest/Person.spec.ts]: why should we test this case? it's the same as previous. You put some data in the constructor and check attributes. It would make sense if you process input somehow.

File: Unittest/Persons.spec.ts [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+  beforeEach(() => {
+    input = [
+      new Person('Zachariah', 'Portugal'),
+      new Person('Lue', 'Nehls'),
+      new Person('Crista', 'Risko'),
+      new Person('Denna', 'Giefer'),
+      new Person('Brandee', 'Hunsberger')
+    ];
+    tmp = new Persons();
```
Comment (fetis) [inline Unittest/Persons.spec.ts]: `tmp` it's too generic variable name, use something more meaningful, like `persons`, the same for `input`, smth like `testPersons` or `personList` would give a more sense.

File: Unittest/Persons.spec.ts [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+  it('should add single Person to the Persons collection', () => {
+    input = new Person('Denna', 'Giefer');
+    let expected = [{firstName: 'Denna', lastName: 'Giefer'}];
+
+    tmp.add(input);
+
+    expect(tmp.items).toEqual(expected);
+  });
+
+  it('should add single Person to the Persons collection with null or undefined arguments', () => {
```
Comment (fetis) [inline Unittest/Persons.spec.ts]: it's the same duplication as from Person.spec.ts

File: Unittest/Persons.spec.ts [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+    expect(tmp.items).toEqual(expected);
+  });
+
+  it('should add an array of Person to the Persons collection', () => {
+    tmp.add(input);
+
+    expect(tmp.items).toEqual(input);
+  });
+
+  it('should add an array of Person to the Persons collection with null or undefined arguments', () => {
```
Comment (fetis) [inline Unittest/Persons.spec.ts]: duplication again.

File: Unittest/Persons.spec.ts [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+      new Person('Brandee', 'Hunsberger')
+    ];
+
+    tmp.add(input);
+
+    expect(tmp.items).toEqual(input);
+  });
+
+  it('should remove a Person from Persons collection', () => {
+    let expected = [
```
Comment (fetis) [inline Unittest/Persons.spec.ts]: if you don't change the variable value it's better to you `const`. Same applies to other tests

File: Unittest/Persons.spec.ts [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+      {firstName: 'Zachariah', lastName: 'Portugal'},
+      {firstName: 'Lue', lastName: 'Nehls'},
+      {firstName: 'Crista', lastName: 'Risko'},
+      {firstName: 'Denna', lastName: 'Giefer'},
+      {firstName: 'Brandee', lastName: 'Hunsberger'}
+    ];
+    tmp.add(input);
+
+    expect(tmp.remove('', null)).toBe(false);
+    expect(tmp.items).toEqual(expected);
```
Comment (fetis) [inline Unittest/Persons.spec.ts]: plz separate execution and check by an empty line

here's TDD presentation I presented recently (can provide you some setup for TDD), page 12 for structure.
https://docs.google.com/presentation/d/1UCxXGLpbdXwUquPbvY7N3kb3ulROZtXi6wZkGVNHJRo/edit?usp=sharing

File: Unittest/Persons.spec.ts [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+    ];
+
+    tmp.add(input);
+    tmp.sortCollection();
+
+    expect(tmp.items).toEqual(expected);
+  });
+
+  it('should call the mock function once and not sort the collection', () => {
+    const f = jest.fn();
```
Comment (fetis) [inline Unittest/Persons.spec.ts]: it's better to call it `spy`

File: Unittest/Persons.spec.ts [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+      {firstName: 'Crista', lastName: 'Risko'}
+    ];
+
+    tmp.add(input);
+    tmp.sortCollection();
+
+    expect(tmp.items).toEqual(expected);
+  });
+
+  it('should call the mock function once and not sort the collection', () => {
```
Comment (fetis) [inline Unittest/Persons.spec.ts]: `and` is a smell that your test does so much. here's better to check only that external function is used. Test description should reflect it, smth like "it should use external sorting if provided"

File: Unittest/Persons.ts [hunk 22]
```diff
@@ -3,34 +3,49 @@ import { Person } from './Person';
...
 
-  constructor() { }
-
   public add(person: Person | Person[]) {
     if(person instanceof Person) {
-      this.collection.push(person);
-    }
-    else {
-      this.collection = this.collection.concat(person);
+      if(this.collection.indexOf(person) === -1) {
```
Comment (fetis) [inline Unittest/Persons.ts]: here you're actually prevent to adding duplicates to the collection, but you don't test this case

you use the same condition twice. what about to move it into a separate method?

File: Unittest/Persons.ts [hunk 23]
```diff
@@ -3,34 +3,49 @@ import { Person } from './Person';
...
   }
 
-  public remove(person: Person): boolean {
+  public remove(firstName: string, lastName: string): boolean {
     let isFound = false;
     for(let i = 0; i < this.collection.length; i++) {
-      if(this.collection[i].compare(person)) {
-        this.collection.splice(i, 1);
-        isFound = true;
+      if(isFound = this.collection[i].firstName === firstName &&
```
Comment (fetis) [inline Unittest/Persons.ts]: this's totally unreadable. the proper way
```
isFound = ...
if (isFound) {...
```

File: Unittest/Person.spec.ts (hunk 14 above)
Comment (TobiasThrebank) [reply Unittest/Person.spec.ts]: I changed it. I think I have a better understanding of what you mean about testing edge cases. This was obviously a poor atempt..

File: Unittest/Persons.spec.ts (hunk 18 above)
Comment (TobiasThrebank) [reply Unittest/Persons.spec.ts]: The linter fixed that easily enough.

File: Unittest/Persons.spec.ts (hunk 21 above)
Comment (TobiasThrebank) [reply Unittest/Persons.spec.ts]: I fixed all tests to only have only one expectation. I hope I haven't overdone it..

File: Unittest/Persons.ts [hunk 24]
```diff
@@ -4,61 +4,72 @@ type SortFunction = (_: Person[]) => void;
...
         }
       }
       n = newn;
-    } while(n > 1);
+    } while (n > 1);
+  }
+
+  private isMember(person: Person): boolean {
+    if (!(person instanceof Person)) {
+      person = new Person(person.firstName, person.lastName);
```
Comment (fetis) [inline Unittest/Persons.ts]: change input parameters is a bad practice. plz use a  variable here

File: Unittest/sortFunction.ts [hunk 25]
```diff
@@ -1,11 +1,15 @@
...
+        const temp = collection[i];
         collection[i] = collection[i + 1];
         collection[i + 1] = temp;
       }
     }
   }
 }
+
+function validateString(str: string): string {
+  return str === null || typeof str === 'undefined' ? '' : str;
```
Comment (fetis) [inline Unittest/sortFunction.ts]: I think it can be shortened to ` str ? str : ''`

File: Unittest/Person.ts [hunk 26]
```diff
@@ -2,7 +2,15 @@ export class Person {
 
   constructor(
     public firstName: string,
-    public lastName: string) { }
+    public lastName: string
+  ) {
+    if (typeof firstName === 'undefined' || firstName === null) {
```
Comment (fetis) [inline Unittest/Person.ts]: it's just `this.firstName = firstName ? firsName : ''`

File: Unittest/Persons.spec.ts [hunk 27]
```diff
@@ -3,122 +3,187 @@ import { Persons } from './Persons';
...
 
-    tmp.add(input);
-    tmp.sortCollection(f);
+  it('should call the mock function without sorting the collection', () => {
+    testCollection.add(testPersons);
+    testCollection.sortCollection(spy);
 
-    expect(f.mock.calls.length).toBe(1);
-    expect(tmp.items).toEqual(input);
+    expect(testCollection.items).toEqual(testPersons);
```
[the remark this answers is not in the document]
Comment (TobiasThrebank) [reply Unittest/Persons.spec.ts]: So theoretically sortFunction.ts and  sortFunction.spec.ts can go aswell?

File: .gitignore [hunk 28]
```diff
@@ -3,3 +3,4 @@ node_modules
 greeter.js
 search_module.js
 Unittest/*.js
+tslint.json
```
[the remark this answers is not in the document]
Comment (fetis) [reply .gitignore]: for many reasons. it's about the teamwork and collaboration. Once you work in a team you need to share same rules for code style. It's also for yourself if you work on different machines or different projects this rules much better to transfer like this file.

File: Unittest/Persons.spec.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (fetis) [reply Unittest/Persons.spec.ts]: If it's part of some external library, then yes we don't care. If it's provided by us, then we still need to test it.

**Comment by fetis** [comment]
I think it's fine now. You can squash & merge, it'll make only 1 commit in master instead of all work commits.
